File: api/app/routers/shopping.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, delete
from app.db.database import get_db
from app.db.models.shopping import ShoppingProduct, ShoppingListItem
from app.schemas.shopping import ShoppingListItemResponse, ShoppingListItemCreate, ShoppingProductResponse, ShoppingProductCreate

router = APIRouter(prefix="/shopping", tags=["shopping"])

# products
@router.get("/products", response_model=list[ShoppingProductResponse])
async def get_products(db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(ShoppingProduct))
    return result.scalars().all()

@router.post("/products", response_model=ShoppingProductResponse)
async def create_product(product: ShoppingProductCreate, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(ShoppingProduct).where(ShoppingProduct.name == product.name))
    existing = result.scalar_one_or_none()
    if existing:
        raise HTTPException(status_code=400, detail="The product is already existing")
    db_product = ShoppingProduct(name=product.name)
    db.add(db_product)
    await db.commit()
    await db.refresh(db_product)
    return db_product

# shopping list
@router.get("/list", response_model=list[ShoppingListItemResponse])
async def get_shopping_list(db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(ShoppingListItem))
    return result.scalars().all()

@router.post("/list", response_model=ShoppingListItemResponse)
async def add_to_list(item: ShoppingListItemCreate, db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(ShoppingProduct).where(ShoppingProduct.id == item.product_id))
    product = result.scalar_one_or_none()
    if not product:
        raise HTTPException(status_code=404, detail="The product does not exist")
    db_item = ShoppingListItem(product_id=item.product_id)
    db.add(db_item)
    await db.commit()
    await db.refresh(db_item)
    return db_item

@router.delete("/list")
async def clear_shopping_list(db: AsyncSession = Depends(get_db)):
    await db.execute(delete(ShoppingListItem))
    await db.commit()
    return {"message": "The shopping list has been cleared"}

File: api/src/routers/shopping.ts
import { Router, Request, Response } from "express";
import { prisma } from "../db/prisma";
import { MESSAGES } from "../messages";

const router = Router();

// Products
router.get("/products", async (_req: Request, res: Response) => {
  const products = await prisma.shoppingProduct.findMany();
  return res.json(products);
});

router.post("/products", async (req: Request, res: Response) => {
  const { name } = req.body;
  const existing = await prisma.shoppingProduct.findUnique({ where: { name } });
  if (existing) return res.status(400).json({ detail: MESSAGES.PRODUCT_ALREADY_EXISTS });

  const product = await prisma.shoppingProduct.create({ data: { name } });
  return res.status(201).json(product);
});

// Shopping list
router.get("/list", async (_req: Request, res: Response) => {
  const items = await prisma.shoppingListItem.findMany({ include: { product: true } });
  return res.json(items);
});

router.post("/list", async (req: Request, res: Response) => {
  const { product_id } = req.body;
  const product = await prisma.shoppingProduct.findUnique({ where: { id: product_id } });
  if (!product) return res.status(404).json({ detail: MESSAGES.PRODUCT_NOT_FOUND });

  const item = await prisma.shoppingListItem.create({
    data: { productId: product_id },
    include: { product: true },
  });
  return res.status(201).json(item);
});

router.delete("/list", async (_req: Request, res: Response) => {
  await prisma.shoppingListItem.deleteMany();
  return res.json({ message: MESSAGES.SHOPPING_LIST_CLEARED });
});

export default router;
